### PR TITLE
feat: positional record destructuring patterns in case (#989)

### LINKED ISSUE #989: feat: support positional record destructuring patterns in `case`
## Context

Issue #834 scoped tuple destructuring patterns in \`case\` to three concerns: (1) tuple patterns, (2) positional record patterns \`Point(a, b)\`, and (3) nested patterns in enum constructor payloads. The tuple-only implementation was completed in the PR for #834. This issue tracks item (2): positional record patterns.

The \`TuplePattern\` foundation added in #834 (AST variant, \`parsePattern\` \`LParen\`-branch, \`emitPatternTest\`/\`emitPatternBindings\` arms, recursive \`isIrrefutable\`) can be reused or extended for this feature.

## Reproduction

```ry
record Point:
    x: int
    y: int

p = Point(3, 4)
case p:
    Point(a, b):   # currently rejected: "expected pattern"
        print(a)
    _:
        print("other")
```

## Expected vs Actual

**Expected:** The \`Point(a, b)\` arm matches and binds \`a = 3\`, \`b = 4\`.
**Actual:** Parse error — positional record pattern syntax is not recognised.

## Discovery timing

During scope-narrowing of #834 implementation (2026-04-16). Decided to split into a separate issue to keep PR size manageable and deliver tuple patterns sooner.

### PR DESCRIPTION
**Pull request by t0k0sh1**
## Summary

Closes #989

- Add `RecordPattern` AST node (`include/ry/ast.hpp`) — `struct RecordPattern { string name; vector<Pattern> elements; }` mirroring `TuplePattern`
- Parser (`src/parser_decl.cpp`): `Ident LParen` branch in `parsePattern` produces `RecordPattern`; `patternHasBinding` recurses over elements
- Codegen (`src/codegen_match.cpp`): `emitPatternTest` validates type/arity and ANDs sub-tests; `emitPatternBindings` extracts fields via `ExtractValue`; `isIrrefutable` treats all-irrefutable record patterns as exhaustive (records have one shape)
- Formatter (`src/formatter.cpp`): prints `Point(a, b)` form
- 6 parser tests, 4 codegen error tests, 16 spec tests (`tests/spec/pattern_record.test.ry`)
- Docs: Record Pattern Matching section added to `docs/reference/control-flow.md`
- Also fixes a latent `TuplePattern` bug: tuple element type signatures were passed as `subjectEnumType` without an enum-only guard, which would crash `valueToString()` via null `CreateGEP` when elements are typed as primitives

## Features

```ry
record Point:
    x: int
    y: int

p = Point(3, 4)

# Binding — positional by declaration order
case p:
    Point(a, b):
        print(a)   # 3
        print(b)   # 4

# Mixed literal / wildcard
case p:
    Point(0, y): print(y)
    _:           print("other")

# Guard
case p:
    Point(a, b) if a > b: print("x bigger")
    Point(a, b):          print("other")

# Nested record-in-record
record Segment:
    start: Point
    end_pt: Point

seg = Segment(Point(1, 2), Point(3, 4))
case seg:
    Segment(Point(x1, _), Point(x2, _)):
        print(x1)   # 1
        print(x2)   # 3

# Expression form
result = case p:
    Point(a, b) => a + b
    _ => -1
```

## Test plan

- [x] `./build/ry_tests` — 1673 tests, all pass
- [x] `./build/ry test -p` — 121 spec files, 0 failures
- [x] ASan + UBSan (`build-asan`) — clean
- [x] TSan (`build-tsan`) — clean

## Summary by CodeRabbit

* **New Features**
  * Positional record destructuring patterns in case arms (e.g., case Point(a, b):) with nested patterns, literals, wildcards, and compile-time arity checks.

* **Bug Fixes**
  * Guarded propagation of enum metadata during pattern bindings to prevent incorrect metadata forwarding for non-enum fields.

* **Documentation**
  * Added reference and knowledge docs covering positional record patterns, syntax, scope, and exhaustiveness rules.

* **Tests**
  * Added parser, codegen, unit, and spec tests covering record patterns, edge cases, and error conditions.

### CODE REVIEW

File: src/codegen_match.cpp [hunk 1]
```diff
@@ -314,6 +317,33 @@ llvm::Value *CodeGen::emitPatternTest(const Pattern &pattern,
...
+            if (sit == struct_types_.end()) {
+                if (enum_types_.count(pat->name))
+                    codegenError("case: '" + pat->name +
+                                 "' is an enum, not a record; use '::' for enum constructor patterns");
+                codegenError("case: unknown record type '" + pat->name + "' in pattern");
+            }
+            // When Ry type metadata is available, verify subject is this record type.
+            if (!subjectEnumType.empty() && subjectEnumType != pat->name)
+                codegenError("case: record pattern '" + pat->name +
+                             "' applied to subject of type '" + subjectEnumType + "'");
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_match.cpp]: Fixed. Added `resolveTypeAlias(pat->name)` → `resolvedName` and used it for both `struct_types_.find()` and `enum_types_.count()` in the RecordPattern arm of `emitPatternTest`. Also normalise `subjectEnumType` with `resolveTypeAlias` before comparing to `resolvedName`, so a type-aliased subject type does not trigger a spurious mismatch error. Applied the same `resolveTypeAlias(pat->name)` lookup in `emitPatternBindings`. Added `CodeGenTest.RecordPatternTypeAlias` to lock the behaviour.

File: src/codegen_match.cpp [hunk 2]
```diff
@@ -367,7 +397,31 @@ void CodeGen::emitPatternBindings(const Pattern &pattern,
...
+                const std::string elemSig = info.fields[i].type->toString();
+                if (!elemSig.empty())
+                    propagateTypeMeta(elemSig, tmp);
+                // Pass elemSig as subjectEnumType only for enum types; primitive and collection
+                // types are already handled by propagateTypeMeta above. Passing "int" or "str"
+                // as subjectEnumType would cause VariablePattern binding to set enum_value_type
+                // to a non-enum name and crash valueToString().
+                const std::string subEnumSig = enum_types_.count(elemSig) ? elemSig : std::string{};
+                emitPatternBindings(pat->elements[i], tmp, elemTy, subEnumSig);
             }
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_match.cpp:433]: Skipping for this PR — this is a pre-existing gap that predates the record pattern feature. The same retain-on-extract is missing in `SomePattern`, `OkPattern`, `ErrPattern`, and `TuplePattern` binding arms as well. Filed as #997 to track the full fix across all pattern arms in a follow-up PR.

**Comment by t0k0sh1** [comment]
@coderabbitai

Addressing the nitpick from your review summary:

**`tests/test_parser.cpp` — add trailing-comma test for `RecordPattern`**

Fixed. Added `ParserTest.RecordPatternTrailingComma` which asserts that `Point(a, b,)` (trailing comma) is accepted without error, matching the existing tuple pattern behaviour. The parser already breaks out of the comma loop when `RParen` follows a comma, so no parser logic change was needed — only the test was missing.

File: src/codegen_match.cpp [hunk 3]
```diff
@@ -367,7 +400,32 @@ void CodeGen::emitPatternBindings(const Pattern &pattern,
                 const std::string elemSig = (i < elemSigs.size()) ? elemSigs[i] : std::string{};
                 if (!elemSig.empty())
                     propagateTypeMeta(elemSig, tmp);
-                emitPatternBindings(pat->elements[i], tmp, elemTy, elemSig);
+                // Guard: only pass elemSig as subjectEnumType when it names an actual enum.
+                // Passing a primitive type name ("int", "str", etc.) would set enum_value_type
+                // to a non-enum name in VariablePattern binding and crash valueToString().
+                const std::string subElemEnumSig = enum_types_.count(elemSig) ? elemSig : std::string{};
+                emitPatternBindings(pat->elements[i], tmp, elemTy, subElemEnumSig);
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_match.cpp:409]: Fixed. Applied `resolveTypeAlias(elemSig)` → `resolvedElemSig` before the `enum_types_.count()` guard in both the `TuplePattern` and `RecordPattern` binding loops. The resolved name is then passed as `subjectEnumType` to the recursive call. This ensures aliased enum field types (e.g. `type MyColor = Color`) are correctly recognised as enums rather than falling through as non-enum primitives.

File: tests/test_codegen_advanced.cpp [hunk 4]
```diff
@@ -940,6 +940,82 @@ TEST_F(CodeGenTest, TuplePatternNonTupleSubject) {
...
+        "type Pt = Point\n"
+        "p = Point(5, 6)\n"
+        "a_out = 0\n"
+        "case p:\n"
+        "    Pt(a, b):\n"
+        "        a_out = a\n"
+        "    _:\n"
+        "        a_out = -1\n";
+    EXPECT_NO_THROW(runSource(src));
+}
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply tests/test_codegen_advanced.cpp:1037]: Fixed. `RecordPatternTypeAlias` now asserts the bound field values with `EXPECT_EQ(runSource(src), "5\n6\n")` — if alias resolution or binding silently falls through, the output diverges and the test fails. Added `RecordPatternSubjectTypeMismatch` to exercise the `resolvedSubject != resolvedName` rejection path with two distinct aliased record types.

**Comment by t0k0sh1** [comment]
@coderabbitai

Addressing the nitpick from your latest review summary:

**`tests/test_parser.cpp` — tighten `RecordPatternTrailingComma` to assert AST shape**

Fixed. `RecordPatternTrailingComma` now parses the source, walks into the `CaseStmt` → `RecordPattern`, asserts `rp.elements.size() == 2` (confirming the trailing comma does not produce a phantom third element), and checks both elements are `VariablePattern`. This makes the test catch structural regressions, not just acceptance.
